src/java/com/rapleaf/jack/queries/GenericQuery.java

```diff
@@ -0,0 +1,149 @@
+package com.rapleaf.jack.queries;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import com.rapleaf.jack.BaseDatabaseConnection;
+import com.rapleaf.jack.ModelField;
+import com.rapleaf.jack.ModelWithId;
+
+public class GenericQuery {
+
+  private final BaseDatabaseConnection dbConnection;
+  private Class<? extends ModelWithId> mainModel;
+  private List<JoinCondition> joinConditions;
+  private List<WhereConstraint> whereConstraints;
+  private List<OrderCriterion> orderCriteria;
+  private Set<ModelField> selectedIModelFields;
+  private Optional<LimitCriterion> limitCondition;
+
+  private GenericQuery(BaseDatabaseConnection dbConnection) {
```

Comment:
Do you plan on adding GROUP BY functionality later? Related to this would be the ability to select functions like MAX, COUNT, etc. Those would probably be some type of generic ModelField that is just a String, not an Enum.
I'm not suggesting that functionality should be part of this version but it's probably important to consider now since it might impact the design of things like ModelField.